### ISSUE TITLE
Allow dynamic lookup of APNS authentication keys via topics

### ISSUE THREAD

**Issue by kdubb** [relayrides/pushy#410]
We have built a platform that provides many applications with a variety of services. For APNS connections we've been using Pushy with great results. 

Since we are a platform with constant uptime, new applications are registered dynamically. Any application that needs to pass information to APNS will provide us with the required keys and topics during registration. This brings me to the configuration challenge we have with Pushy. 

### Problem Outline
---
Pushy, currently, requires that all topics are registered with their associated token before a message can be sent to that topic.  For single application systems, or even multiple application systems with a finite set of applications, this is a simple and practical approach. For our platform this configuration requirement is a bit of a nuisance and doesn't make the best use of resources. 

Here's an example explaining some of the problems.  First, an application is registered with our platform and APNS topics and keys are provided.  The application is used for testing for a few times then left idle for a period of time (say until the next time testing is done).   With Pushy currently we must pre-register the topics and keys for the new application so they are ready when APNS communication is attempted; this could be registered "eagerly" when the application is registered or "lazily" when the first attempt to communicate with APNS about the specified topic is made. If/when the application is unregistered we can unregister the topics and keys from Pushy. So far, no problem, but when the application is left idle and unused it's now consuming resources (memory for keys, topics & CPU as the ApnsRegistry grows in size).  As the number of applications registered grows so does our resource usage.  This is essentially the "memory leak via infinite caching" problem. 

The only solution, currently, is to test the ApnsRegistry to see if the topic is registered then call into Pushy normally; essentially lazy registration. Unfortunately, nothing currently exists to handle the case of unregistering topics/tokens. There is no way to introspect the ApnsRegistry and do cache pruning based on LRU or other methods.

### Proposed Solution
---
Now that the problem description is, hopefully, outlined well here's my suggested solution (this is built upon the code in #404)

Replace `ApnsRegistry` in the `ApnsClientHandler` with a simple interface for locating an `ApnsKey` for a specific topic; e.g. `ApnsKeyLocator`.  That coupled with some method of providing a specific implementation of `ApnsKeyLocator` to the `ApnsClientHandler` would allow users to match implementations to their usage.

I can think of 3 implementations people might need:
* **`ApnsKeyRegistry`**: A **_synchronized_** locator that can be shared between handlers; the registry name implies pre-registered keys. (i.e. the current method)
* **`ApnsConcurrentKeyRegistry`**: A locator that manages one `ApnsKeyRegistry` per handler (eschewing locking). (i.e. the method proposed)
* **`ApnsOnDemandLocator`**: A locator that dynamically looks up keys as they are requested. An implementation based on the Guava `LoadingCache` is imagined.

To be implemented most flexibly you would need two interfaces:
* **`ApnsKeyLocator`**: The interface `ApnsClientHandler` uses to locate keys for specific topics.
* **`ApnsKeyLocatorFactory`**: The interface provided to `ApnsClient` used to create a `ApnsKeyLocator` for each new handler.

**Proposed interfaces for the above:**
```java
interface ApnsKeyLocator {
  /*
    Looks up an ApnsKey for a given topic.
    
    Providing the ApnsClientHandler during locating allows implementations to 
    optimize based on the individual handler (e.g. for concurrency).
   */
  ApnsKey locate(ApnsClientHandler handler, String topic);
}

interface ApnsKeyLocatorFactory {
  /*
    Creates, shares or recycles an ApnsKeyLocator for the handler.

    Not sure these parameters are necessary, or even useful, for actual implementations
    but erring on the side of flexibility and providing as much context as possible.
   */
  ApnsKeyLocator create(ApnsClient client, ApnsClientHandler handler);
}
```

For "compatibility" you could make `ApnsClient` which would work with nothing but `ApnsKeyLocatorFactory` and be the most flexible implementation.  A `SimpleApnsClient` would manage an internal `ApnsKeyRegistery` (imagined above); providing all the same key registration methods `ApnsClient` currently does.

**Comment by kdubb** [comment]
@jchambers As promised I outlined our problem definition and for clarity I consolidated the comments from #404 here as well.

**Comment by jchambers** [comment]
Awesome. Thanks kindly! This definitely clarifies the use case. Thanks for taking the time to write this all out!

My immediate thought is that the next release (v0.10 at the time of this writing) should probably focus on connection pooling, but this seems like a really solid candidate for the following major release. Let me process this a bit; I'll certainly have questions as I work through the ideas here.

**Comment by kdubb** [comment]
@jchambers Sounds good. Glad you are receptive to this. FYI, I have no problem implementing it and providing the PR once a design is laid out.

**Comment by jchambers** [comment]
Much appreciated. Let's get the connection pooling stuff settled out first, though, just in case that dredges up some shocking architectural revelations that would make things difficult here.

**Comment by jchambers** [comment]
Thinking about this a little more, the "do everything we need to do to make everything thread-safe" pull request might actually be the right place to do this. I'll try a couple sketches and see what shakes out.

**Comment by jchambers** [comment]
I tried a thing where we do:

```java
public interface ApnsKeySource<T extends ApnsKey> {
    io.netty.util.concurrent.Future<T> getKeyForTopic(String topic);
}
```

We want this to be asynchronous because it's definitely not a good idea to be blocking on our `EventLoopGroup`'s threads. The challenge, then, is that we can wind up in a situation where the `write` call happens in multiple steps:

1. Start to write a notification
2. Realize we need a not-already-in-memory key to sign the notification
3. Get the key from the key source
4. When we have the key, actually write the notification

This can make the "write and flush" timing really tricky, since `writeAndFlush` seems to (reasonably!) assume that everything gets written in a single call. It seems like the best thing to do here would be to do the asynchronous key stuff *before* calling `write`, but I'm still working through the separation of responsibilities between `ApnsClient` and `ApnsClientHandler` there.

tl;dr: I think this is doable, but need to keep thinking about it.

**Comment by jchambers** [comment]
…flush timing aside, though, this approach does *not* seem to have an adverse affect on performance in preliminary benchmarks.

**Comment by kdubb** [comment]
> because it's definitely not a good idea to be blocking on our EventLoopGroup's threads

Also be mindful that any _sensible_ implementation will use caching if the lookups would take an extended period of time. I think it's safe to assume blocking would be the rare exception; and we certainly would strive to make it be.

**Comment by kdubb** [comment]
I'm kind of warning against pre-optimization here. It seems that synchronous lookups would simplify the implementation of both the `ApnsClientHandler` and `ApnsKeySource`; maybe that's a place to start until benchmarking proves otherwise.

**Comment by jchambers** [comment]
@kdubb As discussed in https://github.com/relayrides/pushy/issues/399#issuecomment-282306488, there's some danger that APNs only supports one "team" per connection. It's unclear whether this is the intended behavior or a bug. We're trying to get some clarity on the issue, but the flow of information has been slow. Depending on the outcome of that conversation, though, the key source thing may (unfortunately) become a non-issue.

**Comment by jchambers** [comment]
Sadly, I think we have to shelve this for now. It turns out that the intended behavior is "one team per connection." I think that means that the whole registry pattern is basically crazy at this point, and we should plan on just binding a single key to a connection.
